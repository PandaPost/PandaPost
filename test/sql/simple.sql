\set ECHO none

\i test/pgxntool/setup.sql

CREATE TEMP TABLE test_cast(
  input 		text
  , test_type 	regtype
  , element_out	text
  , array_out	text
);
INSERT INTO test_cast VALUES
	  ('{t,f,t}',	'boolean[]',	'True',				'[ True False  True]' )
	, ('{1,2,3}',	'smallint[]',	'1',				'[1 2 3]' )
	, ('{1,2,3}',	'int[]',		'1',				'[1 2 3]' )
	, ('{1,2,3}',	'bigint[]',		'1',				'[1 2 3]' )
	, ('{1,2,3}',	'real[]',		'1.0',				'[ 1.  2.  3.]' )
	, ('{1,2,3}',	'float[]',		'1.0',				'[ 1.  2.  3.]' )
	, ('{1,2,3}',	'numeric[]',	'1',				$$[Decimal('1') Decimal('2') Decimal('3')]$$ )
	, ('{a,b,c}',	'text[]',		'a',				$$['a' 'b' 'c']$$ )
;

SELECT plan( (
	1
	+ 6 * (SELECT count(*) FROM test_cast)
		- 2 -- Can't run 2 tests on numeric
)::int );

--SET client_min_messages=debug;
SELECT is(
  str( pg_temp.test_value() )
  , '[  1   2   3 444]'
  , 'Simple test of str'
);

CREATE FUNCTION pg_temp.test_cast(
	i test_cast
) RETURNS SETOF text LANGUAGE plpgsql AS $body$
DECLARE
	element_type CONSTANT regtype := replace( i.test_type::text, '[]', '' );
	n ndarray;
	out text;
BEGIN
	-- ARRAY
	EXECUTE format(
			$$SELECT $1::%s::ndarray$$
			, i.test_type
		)
		INTO n
		USING i.input
	;
	IF element_type <> 'numeric'::regtype THEN
		RETURN NEXT is(
			str(eval('np.'||repr(n)))
			, str(n)
			, 'Test str(eval(repr(n))) = str(ndarray)'
		);
	END IF;

	RETURN NEXT is(
		str(n)
		, i.array_out
		, 'str() output of ' || i.test_type || '::ndarray'
	);

	EXECUTE format(
			$fmt$SELECT is(
				$1::%1$s
				, $2::%1$s
				, 'Cast %1$s to ndarray and back'
			)$fmt$
			, i.test_type
		)
		INTO out
		USING n, i.input
	;
	RETURN NEXT out;


	-- SINGLE ELEMENT
	EXECUTE format(
			$$SELECT ($1::%s)[1]::ndarray$$
			, i.test_type
		)
		INTO n
		USING i.input
	;
	IF element_type <> 'numeric'::regtype THEN
		RETURN NEXT is(
			str(eval('np.'||repr(n)))
			, str(n)
			, 'Test str(eval(repr(n))) = str(ndarray)'
		);
	END IF;

	RETURN NEXT is(
		str(n)
		, i.element_out
		, 'str() output of ' || element_type  || '::ndarray'
	);

	EXECUTE format(
			$fmt$SELECT is(
				$1::%2$s
				, ($2::%1$s)[1]
				, 'Cast %2$s to ndarray and back'
			)$fmt$
			, i.test_type
			, element_type
		)
		INTO out
		USING n, i.input
	;
	RETURN NEXT out;
END
$body$;

SELECT pg_temp.test_cast(t.*) FROM test_cast t;

\i test/pgxntool/finish.sql

-- vi: noexpandtab sw=4 ts=4
