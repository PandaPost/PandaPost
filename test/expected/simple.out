\set ECHO none
1..47
ok 1 - Simple test of str
ok 2 - Test str(eval(repr(n))) = str(ndarray)
ok 3 - str() output of boolean[]::ndarray
ok 4 - Cast boolean[] to ndarray and back
ok 5 - Test str(eval(repr(n))) = str(ndarray)
ok 6 - str() output of boolean::ndarray
ok 7 - Cast boolean to ndarray and back
ok 8 - Test str(eval(repr(n))) = str(ndarray)
ok 9 - str() output of smallint[]::ndarray
ok 10 - Cast smallint[] to ndarray and back
ok 11 - Test str(eval(repr(n))) = str(ndarray)
ok 12 - str() output of smallint::ndarray
ok 13 - Cast smallint to ndarray and back
ok 14 - Test str(eval(repr(n))) = str(ndarray)
ok 15 - str() output of integer[]::ndarray
ok 16 - Cast integer[] to ndarray and back
ok 17 - Test str(eval(repr(n))) = str(ndarray)
ok 18 - str() output of integer::ndarray
ok 19 - Cast integer to ndarray and back
ok 20 - Test str(eval(repr(n))) = str(ndarray)
ok 21 - str() output of bigint[]::ndarray
ok 22 - Cast bigint[] to ndarray and back
ok 23 - Test str(eval(repr(n))) = str(ndarray)
ok 24 - str() output of bigint::ndarray
ok 25 - Cast bigint to ndarray and back
ok 26 - Test str(eval(repr(n))) = str(ndarray)
ok 27 - str() output of real[]::ndarray
ok 28 - Cast real[] to ndarray and back
ok 29 - Test str(eval(repr(n))) = str(ndarray)
ok 30 - str() output of real::ndarray
ok 31 - Cast real to ndarray and back
ok 32 - Test str(eval(repr(n))) = str(ndarray)
ok 33 - str() output of double precision[]::ndarray
ok 34 - Cast double precision[] to ndarray and back
ok 35 - Test str(eval(repr(n))) = str(ndarray)
ok 36 - str() output of double precision::ndarray
ok 37 - Cast double precision to ndarray and back
ok 38 - str() output of numeric[]::ndarray
ok 39 - Cast numeric[] to ndarray and back
ok 40 - str() output of numeric::ndarray
ok 41 - Cast numeric to ndarray and back
ok 42 - Test str(eval(repr(n))) = str(ndarray)
ok 43 - str() output of text[]::ndarray
ok 44 - Cast text[] to ndarray and back
ok 45 - Test str(eval(repr(n))) = str(ndarray)
ok 46 - str() output of text::ndarray
ok 47 - Cast text to ndarray and back
# TRANSACTION INTENTIONALLY LEFT OPEN!
